Prevent insert-only transactions from failing due to concurrent data preserving transactions

zmeir opened the pull request:
This PR prevents insert-only transactions (such as merge operations
with only WHEN NOT MATCHED clauses) from failing due to concurrent data preserving transactions
(i.e. transactions with dataChange=false, such as compaction), as discussed in https://github.com/delta-io/delta/issues/326.  
This ability is controlled by the configuration `concurrentCompactionConflictDetection.enabled`, which is enabled by default.

This change was tested on master with existing and added unit tests in `OptimisticTransactionSuite` and `MergeIntoScalaSuite`. Additionally we have been using a matching implementation of this PR on Delta 0.6.1 in our production process (we're still working with Spark 2.4) for several weeks without any known issues.

In this commit:
* DeltaSQLConf.scala
  * Added configuration `concurrentCompactionConflictDetection.enabled` to control this feature
* OptimisticTransaction.scala
  * Avoid `ConcurrentAppendException` when all new files have been added with `dataChange=false`
  * Avoid `ConcurrentDeleteReadException` when all deleted files have set `dataChange=false`
* OptimisticTransactionSuite.scala
  * Added test cases for transactions with concurrent data rearrangement operation (toggle on/off)
* MergeIntoScalaSuite.scala
  * Added test cases for merge operations with concurrent compaction operation (toggle on/off)

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

jaceklaskowski commented on src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:
```diff
@@ -692,6 +692,14 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
   }
 
+  /**
+   * Used to filter collections of `FileAction`s with `dataChange=false`
```
nit: there's no collections of `FileAction`s so it's misleading.

ZoharMeir replied:
Yes, I realize how how this phrasing doesn’t say exactly what I wanted to describe. My intention was to point out how this function is *used*, and not what it *does*. How about this alternative description:

This function is used to determine if the given `FileAction` should be considered when checking for conflicts based on the value of `dataChange` and the configuration toggle `DeltaSQLConf.CONCURRENT_COMPACTION_CONFLICT_DETECTION`

jaceklaskowski commented on src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:
```diff
@@ -692,6 +692,14 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
+   * Used to filter collections of `FileAction`s with `dataChange=false`
+   * when `DeltaSQLConf.CONCURRENT_COMPACTION_CONFLICT_DETECTION` is enabled.
+   */
+  protected def filterDataChange(f: FileAction): Boolean = {
```
nit: Can you change the name to `isDataChange` or alike. This `filter` can be misleading.

jaceklaskowski replied:
`isDataChangeOrConflict` could work for me too.

ZoharMeir commented:
Hey @jaceklaskowski, are you satisfied with the changes I've made? Can I resolve the open conversations?

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+
+  def createConcurrentCompactionTestTable(location: String): Unit = {
+    spark.range(0)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
```
nit: indent

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+
+  def executeConcurrentCompactionTestInsertOnlyMerge(location: String, i: Int): Unit = {
+    val df = spark.range(i, i + concurrentCompactionTestRangeSize)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
```
nit: formatting / indent

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+  def executeConcurrentCompactionTestInsertOnlyMerge(location: String, i: Int): Unit = {
+    val df = spark.range(i, i + concurrentCompactionTestRangeSize)
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+    val table = io.delta.tables.DeltaTable.forPath(spark, location)
```
nit: add import

ZoharMeir replied:
I used the fully qualified name to be consistent with other tests in this suite, for example: `test("basic scala API")`.   
Should I keep it the same or change it?

jaceklaskowski replied:
I'd at least do `import` just before the line. Just want to get rid of this FQCN here.

ZoharMeir replied:
I think I figured out why the other tests used the FQCN - the import order seems to have messed up the reference to `DeltaTable`.  
I believe I managed to fix that in 109cb55.

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+            .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+    val table = io.delta.tables.DeltaTable.forPath(spark, location)
+    table.alias("t").merge(df.alias("s"), "slowMergeCondition(s.par=t.par AND s.id=t.id)")
+            .whenNotMatched().insertAll()
```
nit: formatting / indent

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+              .coalesce(1)
+              .write.format("delta")
+              .mode("overwrite")
+              .option("dataChange", "false")
```
nit: false as a boolean (not a string)

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,161 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+  class NoException extends Throwable
+
+  def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag]
+  (location: String, executeMergeTest: (String, Int) => Unit): Unit = {
```
nit: input argument formatting (it's hard to read it with all these Scala goodies :))

ZoharMeir replied:
How about this?
```scala
  def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag](
      location: String,
      executeMergeTest: (String, Int) => Unit
  ): Unit = {
```

Or do you prefer this?
```scala
  def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag](
      location: String,
      executeMergeTest: (String, Int) => Unit): Unit = {
```

jaceklaskowski replied:
My preference would be the latter.

```
def runConcurrentCompactionTest[ExpectedExceptionType: ClassTag](
      location: String,
      executeMergeTest: (String, Int) => Unit): Unit = {
```

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala:
```diff
@@ -594,43 +596,343 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
-    withLog(addA_P1 :: addB_P1 :: Nil) { log =>
-      val tx1 = log.startTransaction()
-      tx1.filterFiles()
+  Seq("true", "false").foreach { toggle =>
```
nit: why not boolean values?

ZoharMeir replied:
Because I'm using the `toggle` value in `withSQLConf` which requires pairs of `(String, String)`.  
If you think it's preferable, I'll use boolean values and convert them to String where needed.

jaceklaskowski replied:
It'd make it more type-safe so preferable. Thank you.

jaceklaskowski commented on src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:
```diff
@@ -692,6 +692,15 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
   }
 
+  /**
+   * This function is used to determine if the given `FileAction` should be considered when
```
nit: Start the javadoc with "Determines..."

jaceklaskowski commented on src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:
```diff
@@ -692,6 +692,15 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
+  /**
+   * This function is used to determine if the given `FileAction` should be considered when
+   * checking for conflicts based on the value of `dataChange` and the configuration toggle
+   * `DeltaSQLConf.CONCURRENT_COMPACTION_CONFLICT_DETECTION`.
```
Make it `{@link}`-able

ZoharMeir commented:
Thanks for your comments @jaceklaskowski. I've pushed a new commit which hopefully resolves all of them.  
Let me know if anything else needs to be addressed.

ZoharMeir commented:
Hey @jaceklaskowski, I believe I've addressed all of your comments so far, so if it's ok with you'll I'll go ahead and resolve them.

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,162 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+  val concurrentCompactionTestNumIterations = 10
+
+  def createConcurrentCompactionTestTable(location: String): Unit = {
+    spark.range(0)
```
What's the purpose of this `createConcurrentCompactionTestTable` in general and `spark.range(0)` in particular?

ZoharMeir replied:
This method creates the empty test table to be used in the concurrent compaction test, so that the test iterations could be made simpler, as they can all assume the table already exists and simply perform their merge/compaction operations.

The `spark.range(0)` is just my favorite way of creating an empty DataFrame without having to explicitly specify the schema (it's a single integer column named `id`). Then I just add the `par` column in the next line with the desired number of partitions for the test.

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,162 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+      .range(concurrentCompactionTestNumIterations + concurrentCompactionTestRangeSize - 1)
+      .withColumn("par", $"id" % concurrentCompactionTestPartitions)
+      .orderBy("id")
+    val deltaResultDF = spark.read.format("delta").load(location).orderBy("id")
```
```
import io.delta.tables.DeltaTable
DeltaTable.forPath(location).toDF.orderBy("id")
```

We're in the Delta Lake land so let's use the API as much as possible. The often we use it in tests the more it's tested, no?

Please make sure it's used throughout the PR.

ZoharMeir replied:
Aside from table reads, are there any other places where I should refactor to use the Delta Lake API?

For now, I identified a total of 2 table reads and refactored them both.

jaceklaskowski commented on src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:
```diff
@@ -528,6 +534,162 @@ class MergeIntoScalaSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest {
+
+  def runCompaction(location: String): Unit = {
+    val table = DeltaTable.forPath(spark, location)
+    val lastTableOperation = table.history(1).select("operation").head().getString(0)
```
`as[String].head`

zmeir committed:
Prevent insert-only transactions from failing due to concurrent data preserving transactions.

This commit adds the configuration `concurrentCompactionConflictDetection.enabled`.
When enabled, this configuration will prevent insert-only transactions (such as merge operations
with only WHEN NOT MATCHED clauses) from failing due to concurrent data preserving transactions
(i.e. transactions with dataChange=false, such as compaction).

In this commit:
* DeltaSQLConf.scala
  * Added configuration `concurrentCompactionConflictDetection.enabled` to control this feature
* OptimisticTransaction.scala
  * Avoid `ConcurrentAppendException` when all new files have been added with `dataChange=false`
  * Avoid `ConcurrentDeleteReadException` when all deleted files have set `dataChange=false`
* OptimisticTransactionSuite.scala
  * Added test cases for transactions with concurrent data rearrangement operation (toggle on/off)
* MergeIntoScalaSuite.scala
  * Added test cases for merge operations with concurrent compaction operation (toggle on/off)

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

zmeir committed:
Update OptimisticTransaction.scala

* Renamed `filterDataChange` to the less confusing `isDataChangeOrConflict`
* Clearer documentation for `isDataChangeOrConflict`

zmeir committed:
Styling and formatting changes from PR review

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

zmeir committed:
Minor formatting

zmeir committed:
Changed import order to avoid using FQCN for `DeltaTable`

Moved import of `io.delta.tables._` from global scope to class scope to fix reference to `DeltaTable`:
* Before the change, using `DeltaTable` in the the test class would reference `org.apache.spark.sql.delta.DeltaTable`
* After the change it will reference `io.delta.tables.DeltaTable` as expected

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

zmeir committed:
Refactor table reads in `MergeIntoScalaSuite` to use the Delta API

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

zmeir committed:
Minor formatting

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

jaceklaskowski commented on core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:
```diff
@@ -792,7 +801,10 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
-          addedFilesToCheckForConflicts.toDF(), p :: Nil).as[AddFile].take(1)
+          addedFilesToCheckForConflicts.filter(isDataChangeOrConflict).toDF(),
+          p :: Nil)
+          .as[AddFile]
```
That looks as ugly as before but...you can't do much about that. Sorry for the never-ending begging for change. It's OK.

jaceklaskowski commented on core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:
```diff
@@ -819,7 +831,8 @@ trait OptimisticTransactionImpl extends TransactionalWrite with SQLMetricsReport
       val readFilePaths = readFiles.map(f => f.path -> f.partitionValues).toMap
-      val deleteReadOverlap = removedFiles.find(r => readFilePaths.contains(r.path))
+      val deleteReadOverlap = removedFiles.find(r =>
+        isDataChangeOrConflict(r) & readFilePaths.contains(r.path))
```
Is this single `&` on purpose? (I've never used it before and am curious if `&&` weren't really meant).

ZoharMeir replied:
Wow, great catch! Totally by accident but luckily not logically incorrect, just slightly less efficient:  
https://stackoverflow.com/a/25067787/6877213

Will fix shortly.

jaceklaskowski reviewed (approved):
LGTM (pending auto builds if they're available)

zmeir committed:
Replaced use of strict boolean AND (&) with short-circuit operator (&&)

Signed-off-by: Zohar Meir <zohar.meir@gmail.com>

jaceklaskowski commented:
@tdas @zsxwing @brkyvz Can you have a look at this and hopefully merge? Thank you!

ZoharMeir commented:
> @tdas @zsxwing @brkyvz Can you have a look at this and hopefully merge? Thank you!

Ping :)

zmeir committed:
Merge branch 'master' into pr-concurrent_compaction_conflict_detection

foss-contributor commented:
@zmeir, thank you for your contribution! This pull request has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this pull request or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
